source/Timer.cpp

```diff
@@ -0,0 +1,376 @@
+				proximityCenters.Save(out);
+			}
+			out.Write("proximity settings", proximity, distance);
+		}
+
+		for(const auto &it : actions)
+			it.second.Save(out);
+	}
+	out.EndChild();
+}
+
+
+
+// Calculate the total time to wait, including any random value,
+// and instantiate the triggered action.
+Timer Timer::Instantiate(map<string, string> &subs, const System *origin, int jumps, int64_t payload) const
+{
+	Timer result;
+	result.requireIdle = requireIdle;
+	result.requireUncloaked = requireUncloaked;
+	result.system = system;
+	result.systems = systems;
+	result.proximity = proximity;
+	result.proximityCenter = proximityCenter;
+	result.proximityCenters = proximityCenters;
+	result.closeTo = closeTo;
+	result.resetCondition = resetCondition;
+	result.repeatReset = repeatReset;
+	result.resetFired = resetFired;
+	result.idleMaxSpeed = idleMaxSpeed;
+	result.requirePeaceful = requirePeaceful;
+
+	// Validate all the actions attached to the timer, and if they're *all* valid, instantiate them too
+	string reason;
+	auto ait = actions.begin();
+	for( ; ait != actions.end(); ++ait)
+	{
+		reason = ait->second.Validate();
+		if(!reason.empty())
+			break;
+	}
+	if(ait != actions.end())
+	{
+		Logger::LogError("Instantiation Error: Timer action \"" + TriggerToText(ait->first)
+			+ "\" uses invalid " + std::move(reason));
+		return result;
+	}
+	for(const auto &it : actions)
+		result.actions[it.first] = it.second.Instantiate(subs, origin, jumps, payload);
+
+	// Calculate the random variance to the wait time.
+	result.waitTime = waitTime;
+	if(randomWaitTime > 1)
+		result.waitTime += Random::Int(randomWaitTime);
+
+	// We also build a cache of the matching proximity object(s) for the instantiated Timer.
+	// This avoids having to do all these comparisons every Step().
+	for(const StellarObject &proximityObject : system->Objects())
+		if(proximityObject.HasValidPlanet() && (proximityCenter == proximityObject.GetPlanet() ||
+			proximityCenters.Matches(proximityObject.GetPlanet())))
+			result.proximityCache.insert(&proximityObject);
+
+	return result;
+}
+
+
+
+bool Timer::IsComplete() const
+{
+	return isComplete;
+}
+
+
+
+bool Timer::IsOptional() const
+{
+	return optional;
+}
+
+
+// This method gets called every time the *possible* reset conditions are met,
+// regardless of whether this particular timer is set to reset on them.
+// If it is, then it resets the elapsed time to 0, marks the timer as inactive,
+// and conditionally fires the reset action, if any.
+void Timer::ResetOn(ResetCondition cond, PlayerInfo &player, UI *ui, const Mission &mission)
+{
+	// Check whether this timer actually wants to reset on the given condition.
+	bool reset = cond == resetCondition;
+	reset |= (cond == Timer::ResetCondition::LEAVE_ZONE && resetCondition == Timer::ResetCondition::PAUSE);
+	reset |= (cond == Timer::ResetCondition::LEAVE_SYSTEM && (resetCondition == Timer::ResetCondition::PAUSE
+				|| resetCondition == Timer::ResetCondition::LEAVE_ZONE));
```

Comment:
This seem difficult to read to me. So we reset if the input condition matches the reset condition for this timer, but the input condition and reset condition can differ and we still reset the timer? I feel like calling ResetOn should just always reset the timer, and we only call the function when we need to, instead of calling the function at every possible chance and having the function figure out if it should actually reset.

Reply:
I believe I moved this in here to avoid duplication, as it would need to be checked every time a reset might happen.

I've added some comments to clarify the intention here, which I have pasted below:
```
// If the reset condition being passed in is LEAVE_ZONE (ie, the player is no longer in proximity),
// then we also reset if the timer's reset condition is PAUSE (ie, player no longer meets the criteria).
...
// If the reset condition being passed in is LEAVE_SYSTEM, then we also reset if either of the other two
// conditions is specified for the timer, since we are then necessarily outside the zone, and thus the
// clock is stopped.
```

Please let me know if you still think these tests should be broken back out to the call sites.